### ISSUE TITLE
Installation in PyPy 3.9 fails.

### ISSUE THREAD

**Issue by vtraag** [igraph/python-igraph#527]
**Describe the bug**
Installation in PyPy 3.9 fails. Most likely on PyPy 3.8 also, but I have not yet verified.

<details>
  <summary>Install log</summary>

```
Collecting igraph
  Downloading igraph-0.9.10.tar.gz (3.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.7/3.7 MB 12.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Collecting texttable>=1.6.2
  Using cached texttable-1.6.4-py2.py3-none-any.whl (10 kB)
Building wheels for collected packages: igraph
  Building wheel for igraph (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [1193 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-pypy39
      creating build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/summary.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/operators.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/version.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/statistics.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/matching.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/clustering.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/datatypes.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/utils.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/sparse_matrix.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/configuration.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/formula.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/layout.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/cut.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/__init__.py -> build/lib.linux-x86_64-pypy39/igraph
      creating build/lib.linux-x86_64-pypy39/igraph/app
      copying src/igraph/app/shell.py -> build/lib.linux-x86_64-pypy39/igraph/app
      copying src/igraph/app/__init__.py -> build/lib.linux-x86_64-pypy39/igraph/app
      creating build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/metamagic.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/baseclasses.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/coord.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/vertex.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/colors.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/utils.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/shapes.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/text.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/edge.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/graph.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/__init__.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      creating build/lib.linux-x86_64-pypy39/igraph/remote
      copying src/igraph/remote/gephi.py -> build/lib.linux-x86_64-pypy39/igraph/remote
      copying src/igraph/remote/__init__.py -> build/lib.linux-x86_64-pypy39/igraph/remote
      running build_ext
      running build_c_core
      We are going to build the C core of igraph.
        Source folder: vendor/source/igraph
        Build folder: vendor/build/igraph
        Install folder: vendor/install/igraph
      
      Configuring build...
      -- Setting build type to 'Release' as none was specified.
      -- Version number: 0.9.8
      -- The C compiler identification is GNU 9.4.0
      -- The CXX compiler identification is GNU 9.4.0
      -- Check for working C compiler: /usr/bin/cc
      -- Check for working C compiler: /usr/bin/cc -- works
      -- Detecting C compiler ABI info
      -- Detecting C compiler ABI info - done
      -- Detecting C compile features
      -- Detecting C compile features - done
      -- Check for working CXX compiler: /usr/bin/c++
      -- Check for working CXX compiler: /usr/bin/c++ -- works
      -- Detecting CXX compiler ABI info
      -- Detecting CXX compiler ABI info - done
      -- Detecting CXX compile features
      -- Detecting CXX compile features - done
      -- Performing Test COMPILER_SUPPORTS_NO_VARARGS_FLAG
      -- Performing Test COMPILER_SUPPORTS_NO_VARARGS_FLAG - Success
      -- Performing Test COMPILER_SUPPORTS_NO_UNKNOWN_WARNING_OPTION_FLAG
      -- Performing Test COMPILER_SUPPORTS_NO_UNKNOWN_WARNING_OPTION_FLAG - Success
      -- Looking for pthread.h
      -- Looking for pthread.h - found
      -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
      -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
      -- Looking for pthread_create in pthreads
      -- Looking for pthread_create in pthreads - not found
      -- Looking for pthread_create in pthread
      -- Looking for pthread_create in pthread - found
      -- Found Threads: TRUE
      -- Found FLEX: /usr/bin/flex (found version "2.6.4")
      -- Found BISON: /usr/bin/bison (found version "3.5.1")
      -- Looking for expm1
      -- Looking for expm1 - found
      -- Looking for fmin
      -- Looking for fmin - found
      -- Looking for finite
      -- Looking for finite - found
      -- Looking for isfinite
      -- Looking for isfinite - found
      -- Looking for log2
      -- Looking for log2 - found
      -- Looking for log1p
      -- Looking for log1p - found
      -- Looking for rint
      -- Looking for rint - found
      -- Looking for rintf
      -- Looking for rintf - found
      -- Looking for round
      -- Looking for round - found
      -- Looking for stpcpy
      -- Looking for stpcpy - found
      -- Looking for strcasecmp
      -- Looking for strcasecmp - found
      -- Looking for strdup
      -- Looking for strdup - found
      -- Looking for _stricmp
      -- Looking for _stricmp - not found
      -- Looking for _LIBCPP_VERSION
      -- Looking for _LIBCPP_VERSION - not found
      -- Looking for __GLIBCXX__
      -- Looking for __GLIBCXX__ - found
      -- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
      -- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
      -- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
      -- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
      -- Performing Test COMPILER_HAS_DEPRECATED_ATTR
      -- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success
      -- Found Python3: /home/vtraag/mambaforge/envs/pypy/bin/python3.9 (found version "3.9.12") found components: Interpreter
      --
      -- -----[ Build configuration ]----
      -- Version:                   0.9.8
      -- CMake build type:        Release
      -- Library type:             static
      --
      -- ----------[ Features ]----------
      -- GLPK for optimization:       yes
      -- Reading GraphML files:       yes
      -- Thread-local storage:         no
      -- Link-time optimization:       no
      --
      -- --------[ Dependencies ]--------
      -- ARPACK:                 vendored
      -- BISON:                       yes
      -- BLAS:                   vendored
      -- CXSparse:               vendored
      -- FLEX:                        yes
      -- GLPK:                   vendored
      -- GMP:                    vendored
      -- LAPACK:                 vendored
      -- LibXml2:                     yes
      -- OpenMP:                      yes
      -- PLFIT:                  vendored
      --
      -- -----------[ Testing ]----------
      -- Diff tool:             not found
      -- Sanitizers:                 none
      -- Code coverage:                no
      -- Verify 'finally' stack:       no
      --
      -- --------[ Documentation ]-------
      -- HTML:                        yes
      -- PDF:                         yes
      --
      -- igraph configured successfully.
      --
      -- Configuring done
      -- Generating done
      -- Build files have been written to: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/build/igraph
      Running build...
      Scanning dependencies of target arithchk
      Scanning dependencies of target plfit_vendored
      Scanning dependencies of target gmp_vendored
      Scanning dependencies of target bliss
      Scanning dependencies of target cxsparse_vendored
      [  0%] Building C object vendor/f2c/CMakeFiles/arithchk.dir/arithchk.c.o
      [ ... trunacted because of comment length limits .... ]
      [100%] Built target igraph
      Installing build...
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/lib/libigraph.a
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_statusbar.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_flow.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_decls.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_games.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_community.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_heap_pmt.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_mixing.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_bipartite.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_stack.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_components.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_lapack.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_operators.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_array_pmt.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_sparsemat.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_motifs.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_datatype.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_interrupt.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_graphlets.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_arpack.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_cohesive_blocks.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_eulerian.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_eigen.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_scan.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_paths.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_cliques.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_qsort.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_epidemics.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_scg.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_attributes.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_matrix.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_interface.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_strvector.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_adjlist.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_layout.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_heap.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_constructors.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_vector.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_dqueue_pmt.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_spmatrix.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_structural.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_vector_type.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_cocitation.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_embedding.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_topology.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_foreign.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_transitivity.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_pmt_off.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_neighborhood.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_error.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_lsap.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_complex.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_progress.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_hrg.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_separators.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_types.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_vector_ptr.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_conversion.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_constants.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_stack_pmt.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_matching.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_graphicality.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_coloring.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_microscopic_update.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_iterators.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_matrix_pmt.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_visitor.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_centrality.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_memory.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_nongraph.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_psumtree.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_pmt.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_random.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_array.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_blas.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_vector_pmt.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_dqueue.h
      -- Up-to-date: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_version.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_threading.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/include/igraph/igraph_export.h
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/lib/pkgconfig/igraph.pc
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/lib/cmake/igraph/igraph-config.cmake
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/lib/cmake/igraph/igraph-config-version.cmake
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/lib/cmake/igraph/igraph-targets.cmake
      -- Installing: /tmp/pip-install-9kk87kcy/igraph_f6eb4d72a51b4ff1b017cb0664740013/vendor/install/igraph/lib/cmake/igraph/igraph-targets-release.cmake
      Found igraph as static library in vendor/install/igraph/lib/libigraph.a.
      Build type: dynamic extension with vendored igraph source
      Include path: vendor/install/igraph/include/igraph
      Library path: vendor/install/igraph/lib /usr/local/lib /usr/lib/x86_64-linux-gnu /usr/lib64 /usr/lib /lib64 /lib
      Runtime library path:
      Linked dynamic libraries: m xml2 z gomp pthread
      Linked static libraries: vendor/install/igraph/lib/libigraph.a
      Extra compiler options:
      Extra linker options:
      building 'igraph._igraph' extension
      creating build/temp.linux-x86_64-pypy39
      creating build/temp.linux-x86_64-pypy39/src
      creating build/temp.linux-x86_64-pypy39/src/_igraph
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor/install/igraph/include/igraph -I/home/vtraag/mambaforge/envs/pypy/include/pypy3.9 -c src/_igraph/arpackobject.c -o build/temp.linux-x86_64-pypy39/src/_igraph/arpackobject.o
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor/install/igraph/include/igraph -I/home/vtraag/mambaforge/envs/pypy/include/pypy3.9 -c src/_igraph/attributes.c -o build/temp.linux-x86_64-pypy39/src/_igraph/attributes.o
      In file included from /home/vtraag/mambaforge/envs/pypy/include/pypy3.9/Python.h:58,
                       from src/_igraph/preamble.h:27,
                       from src/_igraph/attributes.h:26,
                       from src/_igraph/attributes.c:25:
      src/_igraph/attributes.c: In function ‘igraphmodule_create_edge_attribute’:
      src/_igraph/attributes.c:236:15: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
        236 |     Py_INCREF(Py_None);
            |               ^~~~~~~
      /home/vtraag/mambaforge/envs/pypy/include/pypy3.9/object.h:32:41: note: in definition of macro ‘_PyObject_CAST’
         32 | #define _PyObject_CAST(op) ((PyObject*)(op))
            |                                         ^~
      src/_igraph/attributes.c:236:5: note: in expansion of macro ‘Py_INCREF’
        236 |     Py_INCREF(Py_None);
            |     ^~~~~~~~~
      src/_igraph/attributes.c:236:15: note: each undeclared identifier is reported only once for each function it appears in
        236 |     Py_INCREF(Py_None);
            |               ^~~~~~~
      /home/vtraag/mambaforge/envs/pypy/include/pypy3.9/object.h:32:41: note: in definition of macro ‘_PyObject_CAST’
         32 | #define _PyObject_CAST(op) ((PyObject*)(op))
            |                                         ^~
      src/_igraph/attributes.c:236:5: note: in expansion of macro ‘Py_INCREF’
        236 |     Py_INCREF(Py_None);
            |     ^~~~~~~~~
      src/_igraph/attributes.c: In function ‘igraphmodule_i_attribute_add_vertices’:
      src/_igraph/attributes.c:500:34: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
        500 |         if (PyList_Append(value, Py_None) == -1) {
            |                                  ^~~~~~~
            |                                  Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_attribute_add_edges’:
      src/_igraph/attributes.c:684:34: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
        684 |         if (PyList_Append(value, Py_None) == -1) {
            |                                  ^~~~~~~
            |                                  Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_ac_first’:
      src/_igraph/attributes.c:1023:72: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1023 |     item = n > 0 ? PyList_GetItem(values, (Py_ssize_t)VECTOR(*v)[0]) : Py_None;
            |                                                                        ^~~~~~~
            |                                                                        Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_ac_random’:
      src/_igraph/attributes.c:1084:14: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1084 |       item = Py_None;
            |              ^~~~~~~
            |              Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_ac_last’:
      src/_igraph/attributes.c:1117:76: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1117 |     item = (n > 0) ? PyList_GetItem(values, (Py_ssize_t)VECTOR(*v)[n-1]) : Py_None;
            |                                                                            ^~~~~~~
            |                                                                            Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_ac_median’:
      src/_igraph/attributes.c:1221:14: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1221 |       item = Py_None;
            |              ^~~~~~~
            |              Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_attribute_get_info’:
      src/_igraph/attributes.c:1498:22: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1498 |             if (o != Py_None && !PyNumber_Check(o))
            |                      ^~~~~~~
            |                      Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_attribute_get_type’:
      src/_igraph/attributes.c:1589:19: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1589 |       if (item != Py_None && !PyNumber_Check(item)) is_numeric=0;
            |                   ^~~~~~~
            |                   Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_get_numeric_graph_attr’:
      src/_igraph/attributes.c:1645:12: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1645 |   if (o == Py_None) {
            |            ^~~~~~~
            |            Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_get_numeric_vertex_attr’:
      src/_igraph/attributes.c:1727:16: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1727 |       if (o != Py_None) {
            |                ^~~~~~~
            |                Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_get_numeric_edge_attr’:
      src/_igraph/attributes.c:1857:16: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1857 |       if (o != Py_None) {
            |                ^~~~~~~
            |                Py_NE
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for igraph
  Running setup.py clean for igraph
Failed to build igraph
Installing collected packages: texttable, igraph
  Running setup.py install for igraph ... error
  error: subprocess-exited-with-error
  
  × Running setup.py install for igraph did not run successfully.
  │ exit code: 1
  ╰─> [136 lines of output]
      running install
      /home/vtraag/mambaforge/envs/pypy/lib/pypy3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-pypy39
      creating build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/summary.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/operators.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/version.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/statistics.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/matching.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/clustering.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/datatypes.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/utils.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/sparse_matrix.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/configuration.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/formula.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/layout.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/cut.py -> build/lib.linux-x86_64-pypy39/igraph
      copying src/igraph/__init__.py -> build/lib.linux-x86_64-pypy39/igraph
      creating build/lib.linux-x86_64-pypy39/igraph/app
      copying src/igraph/app/shell.py -> build/lib.linux-x86_64-pypy39/igraph/app
      copying src/igraph/app/__init__.py -> build/lib.linux-x86_64-pypy39/igraph/app
      creating build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/metamagic.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/baseclasses.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/coord.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/vertex.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/colors.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/utils.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/shapes.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/text.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/edge.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/graph.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      copying src/igraph/drawing/__init__.py -> build/lib.linux-x86_64-pypy39/igraph/drawing
      creating build/lib.linux-x86_64-pypy39/igraph/remote
      copying src/igraph/remote/gephi.py -> build/lib.linux-x86_64-pypy39/igraph/remote
      copying src/igraph/remote/__init__.py -> build/lib.linux-x86_64-pypy39/igraph/remote
      running build_ext
      running build_c_core
      Found igraph as static library in vendor/install/igraph/lib/libigraph.a.
      Build type: dynamic extension with vendored igraph source
      Include path: vendor/install/igraph/include/igraph
      Library path: vendor/install/igraph/lib /usr/local/lib /usr/lib/x86_64-linux-gnu /usr/lib64 /usr/lib /lib64 /lib
      Runtime library path:
      Linked dynamic libraries: m xml2 z gomp pthread
      Linked static libraries: vendor/install/igraph/lib/libigraph.a
      Extra compiler options:
      Extra linker options:
      building 'igraph._igraph' extension
      creating build/temp.linux-x86_64-pypy39
      creating build/temp.linux-x86_64-pypy39/src
      creating build/temp.linux-x86_64-pypy39/src/_igraph
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor/install/igraph/include/igraph -I/home/vtraag/mambaforge/envs/pypy/include/pypy3.9 -c src/_igraph/arpackobject.c -o build/temp.linux-x86_64-pypy39/src/_igraph/arpackobject.o
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor/install/igraph/include/igraph -I/home/vtraag/mambaforge/envs/pypy/include/pypy3.9 -c src/_igraph/attributes.c -o build/temp.linux-x86_64-pypy39/src/_igraph/attributes.o
      In file included from /home/vtraag/mambaforge/envs/pypy/include/pypy3.9/Python.h:58,
                       from src/_igraph/preamble.h:27,
                       from src/_igraph/attributes.h:26,
                       from src/_igraph/attributes.c:25:
      src/_igraph/attributes.c: In function ‘igraphmodule_create_edge_attribute’:
      src/_igraph/attributes.c:236:15: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
        236 |     Py_INCREF(Py_None);
            |               ^~~~~~~
      /home/vtraag/mambaforge/envs/pypy/include/pypy3.9/object.h:32:41: note: in definition of macro ‘_PyObject_CAST’
         32 | #define _PyObject_CAST(op) ((PyObject*)(op))
            |                                         ^~
      src/_igraph/attributes.c:236:5: note: in expansion of macro ‘Py_INCREF’
        236 |     Py_INCREF(Py_None);
            |     ^~~~~~~~~
      src/_igraph/attributes.c:236:15: note: each undeclared identifier is reported only once for each function it appears in
        236 |     Py_INCREF(Py_None);
            |               ^~~~~~~
      /home/vtraag/mambaforge/envs/pypy/include/pypy3.9/object.h:32:41: note: in definition of macro ‘_PyObject_CAST’
         32 | #define _PyObject_CAST(op) ((PyObject*)(op))
            |                                         ^~
      src/_igraph/attributes.c:236:5: note: in expansion of macro ‘Py_INCREF’
        236 |     Py_INCREF(Py_None);
            |     ^~~~~~~~~
      src/_igraph/attributes.c: In function ‘igraphmodule_i_attribute_add_vertices’:
      src/_igraph/attributes.c:500:34: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
        500 |         if (PyList_Append(value, Py_None) == -1) {
            |                                  ^~~~~~~
            |                                  Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_attribute_add_edges’:
      src/_igraph/attributes.c:684:34: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
        684 |         if (PyList_Append(value, Py_None) == -1) {
            |                                  ^~~~~~~
            |                                  Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_ac_first’:
      src/_igraph/attributes.c:1023:72: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1023 |     item = n > 0 ? PyList_GetItem(values, (Py_ssize_t)VECTOR(*v)[0]) : Py_None;
            |                                                                        ^~~~~~~
            |                                                                        Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_ac_random’:
      src/_igraph/attributes.c:1084:14: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1084 |       item = Py_None;
            |              ^~~~~~~
            |              Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_ac_last’:
      src/_igraph/attributes.c:1117:76: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1117 |     item = (n > 0) ? PyList_GetItem(values, (Py_ssize_t)VECTOR(*v)[n-1]) : Py_None;
            |                                                                            ^~~~~~~
            |                                                                            Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_ac_median’:
      src/_igraph/attributes.c:1221:14: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1221 |       item = Py_None;
            |              ^~~~~~~
            |              Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_attribute_get_info’:
      src/_igraph/attributes.c:1498:22: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1498 |             if (o != Py_None && !PyNumber_Check(o))
            |                      ^~~~~~~
            |                      Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_attribute_get_type’:
      src/_igraph/attributes.c:1589:19: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1589 |       if (item != Py_None && !PyNumber_Check(item)) is_numeric=0;
            |                   ^~~~~~~
            |                   Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_get_numeric_graph_attr’:
      src/_igraph/attributes.c:1645:12: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1645 |   if (o == Py_None) {
            |            ^~~~~~~
            |            Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_get_numeric_vertex_attr’:
      src/_igraph/attributes.c:1727:16: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1727 |       if (o != Py_None) {
            |                ^~~~~~~
            |                Py_NE
      src/_igraph/attributes.c: In function ‘igraphmodule_i_get_numeric_edge_attr’:
      src/_igraph/attributes.c:1857:16: error: ‘Py_None’ undeclared (first use in this function); did you mean ‘Py_NE’?
       1857 |       if (o != Py_None) {
            |                ^~~~~~~
            |                Py_NE
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> igraph

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
```

</details>

**To reproduce**

There are currently no binary wheels provided for PyPy > 3.7, so you can simply install from source using `pypy3 -m pip install igraph`. I have created a PyPy environment via conda first:

```
conda config --set channel_priority strict
conda create -n pypy pypy
conda activate pypy
```

**Version information**

0.9.10 from PyPI.

**Comment by ntamas** [comment]
That's interesting; `Py_None` is a fundamental part of Python's C API so I don't understand how it could be missing on PyPy and why it was present with PyPy <= 3.7.

**Comment by vtraag** [comment]
Yeah, I also didn't quite understand it, I'm not sure where it comes from exactly, `Python.h` does seem to be correctly included. I arrived at this problem when trying to use `cibuildwheel` but it wasn't able to install `igraph` for PyPy.

**Comment by vtraag** [comment]
Perhaps something related to `Py_LIMITED_API`?

**Comment by vtraag** [comment]
Indeed, in `object.h` it seems to say

```C
#ifndef Py_LIMITED_API
...
#define Py_None (&_Py_NoneStruct)

#endif
```

I'm not sure what's up with that?

**Comment by ntamas** [comment]
That might be it, although note that the code compiles correctly for CPython so in this case this could be a bug in CPython.

The role of `Py_LIMITED_API` is to commit ourselves to use only the stable API of CPython that was valid in a particular version. This will allow CPython in the long run to compile a _single_ wheel for _all_ Python versions greater than the one we've committed ourselves to.

CPython seems to have no problems with using `Py_None` when `Py_LIMITED_API` is defined. On the other hand, `Py_None` is not officially documented as [part of the stable API](https://docs.python.org/3/c-api/stable.html) so we should probably not use it. The workaround for the time being is to remove `#define Py_LIMITED_API` from everywhere you see it.

**Comment by vtraag** [comment]
Rather unexpected that `Py_None` is not part of the limited stable API, wouldn't it? We can simply remove `Py_LIMITED_API` everywhere of course, but it would also defeat the purpose a bit, it seems.

**Comment by ntamas** [comment]
It's just temporary because the current version is not committed to the stable API yet; the inclusion of `Py_LIMITED_API` was an oversight on my end (it could have accidentally been merged from `develop` back into `master`). We'll need to fix this properly on the `develop` branch but the `master` branch should be okay without it.

The reason why `Py_None` is not part of the stable API is because it is a macro that resolves to the address of a hidden `_Py_NoneStruct` object. The stable API must have some kind of a replacement for it that is a proper function, I just need to find it.

**Comment by vtraag** [comment]
Right, OK. Most of the checks you seem to perform can be replaced by [`Py_IsNone`](https://docs.python.org/3/c-api/structures.html#c.Py_IsNone) it seems. But appending a new `Py_None` value seems rather tricky without it being defined directly.

**Comment by vtraag** [comment]
Returning a None value should be possible to replace by `Py_RETURN_NONE`, I guess?

**Comment by ntamas** [comment]
`Py_RETURN_NONE` is not part of the stable API either, according to [this list](https://docs.python.org/3/c-api/stable.html#contents-of-limited-api). `Py_BuildValue("s", NULL)` is an alternative but maybe there's a simpler option.

**Comment by xylar** [comment]
> Most likely on PyPy 3.8 also, but I have not yet verified.

For what it's worth, this problem seems only to be happening for PyPy 3.9 on conda-forge:
https://github.com/conda-forge/python-igraph-feedstock/pull/56

**Comment by ntamas** [comment]
FYI, `master` now builds successfully on PyPy 3.9 after https://github.com/igraph/python-igraph/commit/806cf9f1ac4272ed3c5e7cb001b4b9c4d7f95ac9

`develop` fails, but that's expected, we'll deal with the `Py_None` issue there.

**Comment by ntamas** [comment]
I'm inclined to bend the rules a bit and take the most recent PyPy 3.8 and 3.9 wheels from the `master` CI build and upload them to PyPI for version 0.9.10 (even though technically they are one commit _after_ 0.9.10). Any objections?

Also, since the builds in `master` are fixed, I'm closing this and opening another issue for dealing with `Py_None` in the `develop` branch.

**Comment by vtraag** [comment]
> I'm inclined to bend the rules a bit and take the most recent PyPy 3.8 and 3.9 wheels from the `master` CI build and upload them to PyPI for version 0.9.10 (even though technically they are one commit _after_ 0.9.10). Any objections?

Nope, no objections. For building in `conda-forge` for PyPy we've now also patched the release until a proper new release is made (see https://github.com/conda-forge/python-igraph-feedstock/pull/56).
